In [1]:
import warnings
warnings.filterwarnings('ignore')

# 1.	Read the data and remove all null or empty values.

In [3]:
from csv import reader

# Initialize an empty list to store transactions
groceries = []

# Read the CSV file and filter out empty strings
with open('Dataset_Day15.csv', 'r') as csvfile:
    read = reader(csvfile)
    for row in read:
        # Filter out empty strings
        row_cleaned = [item for item in row if item]
        groceries.append(row_cleaned)

In [4]:
groceries[0:3]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney']]

### 2.	Transform the data using TransactionEncoder to perform Market Basket Analysis.

In [5]:
!pip install mlxtend

In [6]:
from mlxtend.preprocessing import TransactionEncoder

encoder = TransactionEncoder()

transactions = encoder.fit(groceries).transform(groceries)
transactions
import pandas as pd

itemsets = pd.DataFrame(transactions, columns = encoder.columns_)

In [7]:
display(itemsets.head())
display(itemsets.info())

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Columns: 120 entries,  asparagus to zucchini
dtypes: bool(120)
memory usage: 879.1 KB


None

### 3.	 Use min_support = 0.02 to find frequent itemsets.

In [8]:
from mlxtend.frequent_patterns import apriori
# Find frequent itemsets with a minimum support of 0.02
frequent_itemsets = apriori(itemsets, min_support = 0.02, use_colnames = True)
frequent_itemsets

support                           itemsets
0    0.020397                          (almonds)
1    0.033329                          (avocado)
2    0.033729                         (brownies)
3    0.087188                          (burgers)
4    0.030129                           (butter)
..        ...                                ...
98   0.020131  (mineral water, whole wheat rice)
99   0.022930             (spaghetti, olive oil)
100  0.025197              (spaghetti, pancakes)
101  0.021197                (spaghetti, shrimp)
102  0.020931              (spaghetti, tomatoes)

[103 rows x 2 columns]

In [9]:
# Sort the frequent itemsets by support in descending order
frequent_itemsets.sort_values('support', ascending = False)

support                           itemsets
34  0.238368                    (mineral water)
13  0.179709                             (eggs)
44  0.174110                        (spaghetti)
17  0.170911                     (french fries)
9   0.163845                        (chocolate)
..       ...                                ...
0   0.020397                          (almonds)
80  0.020264   (mineral water, frozen smoothie)
98  0.020131  (mineral water, whole wheat rice)
78  0.020131           (pancakes, french fries)
67  0.020131       (mineral water, cooking oil)

[103 rows x 2 columns]

### 4.	Use the frequent itemsets to create association rules (take min_threshold as 15%) and evaluate the rules for the following metrics

In [10]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.15)
rules

antecedents      consequents  antecedent support  \
0            (burgers)           (eggs)            0.087188   
1               (eggs)        (burgers)            0.179709   
2            (burgers)   (french fries)            0.087188   
3            (burgers)  (mineral water)            0.087188   
4            (burgers)      (spaghetti)            0.087188   
..                 ...              ...                 ...   
66  (whole wheat rice)  (mineral water)            0.058526   
67         (olive oil)      (spaghetti)            0.065858   
68          (pancakes)      (spaghetti)            0.095054   
69            (shrimp)      (spaghetti)            0.071457   
70          (tomatoes)      (spaghetti)            0.068391   

    consequent support   support  confidence      lift  leverage  conviction  \
0             0.179709  0.028796    0.330275  1.837830  0.013128    1.224818   
1             0.087188  0.028796    0.160237  1.837830  0.013128    1.086988   
2             0.170911  0.021997    0.252294  1.476173  0.007096    1.108844   
3             0.238368  0.024397    0.279817  1.173883  0.003614    1.057552   
4             0.174110  0.021464    0.246177  1.413918  0.006283    1.095602   
..                 ...       ...         ...       ...       ...         ...   
66            0.238368  0.020131    0.343964  1.442993  0.006180    1.160960   
67            0.174110  0.022930    0.348178  1.999758  0.011464    1.267048   
68            0.174110  0.025197    0.265077  1.522468  0.008647    1.123778   
69            0.174110  0.021197    0.296642  1.703760  0.008756    1.174209   
70            0.174110  0.020931    0.306043  1.757755  0.009023    1.190117   

    zhangs_metric  
0        0.499424  
1        0.555754  
2        0.353384  
3        0.162275  
4        0.320707  
..            ...  
66       0.326080  
67       0.535186  
68       0.379218  
69       0.444850  
70       0.462740  

[71 rows x 10 columns]

In [11]:
rules.describe()

antecedent support  consequent support    support  confidence  \
count           71.000000           71.000000  71.000000   71.000000   
mean             0.131766            0.179854   0.031224    0.255879   
std              0.050738            0.043768   0.009823    0.076048   
min              0.050527            0.087188   0.020131    0.152374   
25%              0.095054            0.163845   0.023530    0.201295   
50%              0.129583            0.174110   0.028530    0.234735   
75%              0.174110            0.238368   0.036062    0.316794   
max              0.238368            0.238368   0.059725    0.456464   

            lift   leverage  conviction  zhangs_metric  
count  71.000000  71.000000   71.000000      71.000000  
mean    1.440443   0.008550    1.110022       0.316343  
std     0.300748   0.005431    0.089046       0.170015  
min     0.827912  -0.007011    0.948894      -0.200452  
25%     1.208565   0.005460    1.048105       0.210909  
50%     1.421397   0.008090    1.089850       0.329550  
75%     1.624075   0.011267    1.158047       0.442659  
max     2.291162   0.022088    1.401255       0.682343

# Lift
# Mean: 1.44
# Range: 0.83 to 2.29
# Key Insight: The average lift of 1.44 indicates that, on average, the antecedent increases the likelihood of the consequent by 44% compared to its typical occurrence. Values span from 0.83 (indicating a weak or potentially negative association) to 2.29 (indicating a strong positive association).

# Leverage
# Mean: 0.009
# Range: -0.007 to 0.022
# Key Insight: The average leverage of 0.009 suggests a slight positive association between antecedents and consequents. The range shows that most leverage values are very close to zero, meaning the actual frequency of the rules is close to what would be expected if antecedents and consequents were independent.

# Conviction
# Mean: 1.11
# Range: 0.95 to 1.40
# Key Insight: An average conviction of 1.11 indicates a slight predictive power of the rules. 
# The values range from 0.95 (indicating weak prediction strength) to 1.40 (moderate prediction strength), suggesting that the rules are somewhat reliable but not very strong predictors.

# Zhang’s Metric
# Mean: 0.316
# Range: -0.200 to 0.682
# Key Insight: The average Zhang’s metric of 0.316 suggests that, generally, removing the antecedent has a positive impact on the consequent's occurrence. 
#Negative values indicate a negative correlation, while positive values indicate a positive correlation.

# Support
# Mean: 0.031
# Range: 0.020 to 0.060
# Key Insight: The average support of 0.031 means that the rules cover about 3.1% of the transactions. 
#The range indicates low to moderate coverage, with rules appearing in 2% to 6% of the transactions.

# Confidence
# Mean: 0.256
# Range: 0.152 to 0.456
# Key Insight: The average confidence of 0.256 shows that the antecedent predicts the consequent 25.6% of the time. 
#The range from 15.2% to 45.6% indicates varying levels of predictiveness, from weak to moderate.

# Antecedent Support
# Mean: 0.132
# Range: 0.051 to 0.238
# Key Insight: The average antecedent support of 0.132 means that the antecedents appear in 13.2% of transactions. 
#The range shows that some antecedents are relatively uncommon (5.1%) while others are more frequent (23.8%).

# Consequent Support
# Mean: 0.180
# Range: 0.087 to 0.238
# Key Insight: The average consequent support of 0.180 indicates that the consequents appear in 18.0% of transactions. 
#The range from 8.7% to 23.8% suggests a similar frequency of consequents, with some being more common than others.

### a.	Find top 5 antecedent -> consequent rules based on ‘leverage’, ‘lift’ each. Explain your findings.

In [13]:
rules.sort_values('lift',ascending=False).head()

antecedents          consequents  antecedent support  \
52    (spaghetti)        (ground meat)            0.174110   
53  (ground meat)          (spaghetti)            0.098254   
67    (olive oil)          (spaghetti)            0.065858   
62         (soup)      (mineral water)            0.050527   
41         (milk)  (frozen vegetables)            0.129583   

    consequent support   support  confidence      lift  leverage  conviction  \
52            0.098254  0.039195    0.225115  2.291162  0.022088    1.163716   
53            0.174110  0.039195    0.398915  2.291162  0.022088    1.373997   
67            0.174110  0.022930    0.348178  1.999758  0.011464    1.267048   
62            0.238368  0.023064    0.456464  1.914955  0.011020    1.401255   
41            0.095321  0.023597    0.182099  1.910382  0.011245    1.106099   

    zhangs_metric  
52       0.682343  
53       0.624943  
67       0.535186  
62       0.503221  
41       0.547490

# Observations for Top 5 Rules Based on ‘Lift’
# spaghetti -> ground meat: Lift = 2.29
# Customers who buy spaghetti are 2.29 times more likely to also buy ground meat.

# A lift value above 1 indicates a strong positive association.

# ground meat -> spaghetti: Lift = 2.29
# Customers who buy ground meat are 2.29 times more likely to also buy spaghetti.

# This mutual high lift value suggests a strong relationship between these items.

# olive oil -> spaghetti: Lift = 2.00
# Customers who buy olive oil are 2.00 times more likely to also buy spaghetti.

# This indicates a strong association between olive oil and spaghetti.

# soup -> mineral water: Lift = 1.91
# Customers who buy soup are 1.91 times more likely to also buy mineral water.

# This shows a significant positive relationship between soup and mineral water.

# milk -> frozen vegetables: Lift = 1.91
# Customers who buy milk are 1.91 times more likely to also buy frozen vegetables.

# A lift value above 1 signifies a strong association between milk and frozen vegetables.

In [15]:
rules.sort_values('leverage',ascending=False).head()

antecedents      consequents  antecedent support  consequent support  \
52      (spaghetti)    (ground meat)            0.174110            0.098254   
53    (ground meat)      (spaghetti)            0.098254            0.174110   
64  (mineral water)      (spaghetti)            0.238368            0.174110   
63      (spaghetti)  (mineral water)            0.174110            0.238368   
50  (mineral water)    (ground meat)            0.238368            0.098254   

     support  confidence      lift  leverage  conviction  zhangs_metric  
52  0.039195    0.225115  2.291162  0.022088    1.163716       0.682343  
53  0.039195    0.398915  2.291162  0.022088    1.373997       0.624943  
64  0.059725    0.250559  1.439085  0.018223    1.102008       0.400606  
63  0.059725    0.343032  1.439085  0.018223    1.159314       0.369437  
50  0.040928    0.171700  1.747522  0.017507    1.088672       0.561638

# Observations for Top 5 Rules Based on ‘Leverage’
# spaghetti -> ground meat: Leverage = 0.022
# Customers who bought spaghetti are likely to also purchase ground meat.

# This strong co-occurrence is backed by a lift score of 2.29, indicating a significant association.

# ground meat -> spaghetti: Leverage = 0.022
# Customers who bought ground meat are also likely to purchase spaghetti.

# This rule mirrors the strong association seen in the first rule, with a lift of 2.29.

# olive oil -> spaghetti: Leverage = 0.011
# Customers who bought olive oil are somewhat likely to also purchase spaghetti.

# The lift of 2.00 suggests a significant relationship, though the leverage is lower.

# soup -> mineral water: Leverage = 0.011
# The rule indicates that customers buying soup are likely to also buy mineral water.

# A lift of 1.91 shows a notable association.

# milk -> frozen vegetables: Leverage = 0.011
# Customers who bought milk are likely to also purchase frozen vegetables.

# Despite a high lift of 1.91, the leverage value is lower, suggesting a weaker but still significant association.

### b.	Find top 2 & bottom 2 antecedent -> consequent rules based on ‘zhang’s metric’. Explain your findings.

In [17]:
rules.sort_values('zhangs_metric',ascending=False).head(2)


antecedents    consequents  antecedent support  consequent support  \
52    (spaghetti)  (ground meat)            0.174110            0.098254   
53  (ground meat)    (spaghetti)            0.098254            0.174110   

     support  confidence      lift  leverage  conviction  zhangs_metric  
52  0.039195    0.225115  2.291162  0.022088    1.163716       0.682343  
53  0.039195    0.398915  2.291162  0.022088    1.373997       0.624943

# Rule 1: Spaghetti -> Ground meat: Zhang’s Metric = 0.68
# Support: 0.039195 (3.92% of transactions have this rule)
# Confidence: 0.225115 (22.51% of transactions with spaghetti also have ground meat)
# Lift: 2.291162 (Ground meat is 2.29 times more likely to be bought with spaghetti)
# Zhang's Metric: 0.682343 (Strong positive association)

# Rule 2: Ground meat -> Spaghetti: Zhang’s Metric = 0.62
# Support: 0.039195 (3.92% of transactions have this rule)
# Confidence: 0.398915 (39.89% of transactions with ground meat also have spaghetti)
# Lift: 2.291162 (Spaghetti is 2.29 times more likely to be bought with ground meat)
# Zhang's Metric: 0.624943 (Strong positive association)
# These rules show a strong positive association between spaghetti and ground meat, indicating that they are frequently bought together.

In [19]:
rules.sort_values('zhangs_metric',ascending=True).head(2)

antecedents      consequents  antecedent support  consequent support  \
35  (french fries)  (mineral water)            0.170911            0.238368   
37     (spaghetti)   (french fries)            0.174110            0.170911   

     support  confidence      lift  leverage  conviction  zhangs_metric  
35  0.033729    0.197348  0.827912 -0.007011    0.948894      -0.200452  
37  0.027596    0.158499  0.927381 -0.002161    0.985251      -0.086602

# Antecedents and Consequents:
# The first rule indicates that having 'french fries' as an antecedent is associated with 'mineral water' as a consequent.
# The second rule indicates that having 'spaghetti' as an antecedent is associated with 'french fries' as a consequent.

# Support:
# The support for 'french fries' leading to 'mineral water' is 0.033729, meaning that this rule appears in 3.37% of the transactions.
# The support for 'spaghetti' leading to 'french fries' is 0.027596, meaning that this rule appears in 2.76% of the transactions.

# Confidence:
# The confidence for 'french fries' leading to 'mineral water' is 0.197348, meaning that 19.73% of the transactions containing 'french fries' also contain 'mineral water'.
# The confidence for 'spaghetti' leading to 'french fries' is 0.158499, meaning that 15.85% of the transactions containing 'spaghetti' also contain 'french fries'.

# Lift:
# The lift for 'french fries' leading to 'mineral water' is 0.827912, indicating that 'mineral water' is less likely to be bought when 'french fries' are bought compared to its overall purchase frequency.
# The lift for 'spaghetti' leading to 'french fries' is 0.927381, indicating that 'french fries' are also less likely to be bought when 'spaghetti' is bought compared to its overall purchase frequency.

# Leverage:
# The leverage for 'french fries' leading to 'mineral water' is -0.007011, indicating a slight negative association between 'french fries' and 'mineral water'.
# The leverage for 'spaghetti' leading to 'french fries' is -0.002161, indicating a slight negative association between 'spaghetti' and 'french fries'.

# Conviction:
# The conviction for 'french fries' leading to 'mineral water' is 0.948894, which measures the degree of implication or dependence between 'french fries' and 'mineral water'.
# The conviction for 'spaghetti' leading to 'french fries' is 0.985251, which measures the degree of implication or dependence between 'spaghetti' and 'french fries'.

# Zhang's Metric:
# Both rules have negative Zhang's metric values, indicating a negative correlation between the antecedents and consequents.
# The first rule ('french fries' -> 'mineral water') has a Zhang's metric of -0.200452.
# The second rule ('spaghetti' -> 'french fries') has a Zhang's metric of -0.086602.